src/Rules/Actions.hs

```diff
@@ -126,6 +128,18 @@ copyDirectory source target = do
     putProgressInfo $ renderAction "Copy directory" source target
     quietly $ cmd cmdEcho ["cp", "-r", source, target]
 
+-- | Copy the content of the source directory into the target directory. Only
+-- the files and directories for which the predicate returns True are copied.
+copyDirectoryContent :: (FilePath -> IO Bool) -> FilePath -> FilePath -> Action ()
+copyDirectoryContent test source target = do
+    putProgressInfo $ renderAction "Copy directory" source target
+    liftIO $ X.listFilesInside test' source >>= mapM_ cp
+  where
+    target' a = target -/- fromJust (stripPrefix source a)
+    test' a = ifM (test a) (mkdir a >> return True) (return False)
+    mkdir a = IO.createDirectoryIfMissing True $ target' a
+    cp a = whenM (test a) $ IO.copyFile a $ target' a
```

Comment:
@ndmitchell I think we actually don't need a predicate. We only need to exclude certain `FilePattern`'s. So, it's like going from
```haskell
getDirectoryFiles :: FilePath -> [FilePattern] -> Action [FilePath]
getDirectoryFiles dir include = do ...
```
to
```haskell
getDirectoryFiles' :: FilePath -> [FilePattern] -> [FilePattern] -> Action [FilePath]
getDirectoryFiles' dir include exclude = do ...
```
We can certainly try to implement it in Hadrian directly and see how it goes.